aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -52,13 +57,92 @@ static void index_select_add(const Tensor &select_indices,
   auto output_stride0 = output.stride(0);
   auto output_stride1 = output.stride(1);
 
-  for (int64_t i = 0; i < numel; i++) {
-    THBlas_axpy<T>(ddim, 1,
-            src_data + src_stride0 * select_indices_data[i], src_stride1,
-            output_data + output_stride0 * add_indices_data[i], output_stride1);
+  auto accessor = offsets.accessor<int64_t, 1>();
+  std::vector<int> lengths(offsets.numel());
+  int64_t lower = accessor[0];
+  for (size_t i = 1; i < offsets.numel(); ++i) {
+    if (i == offsets.numel() - 1) {
+      lengths[i] = add_indices.size(0) - lower;
+    } else {
+      lengths[i] = accessor[i] - lower;
+    }
+    lower = accessor[i];
+  }
+
+  // We specialize the case where both the src and output are contiguous in the
+  // last dimension to give the compiler a better chance to vectorize.
+  if (src_stride1 == 1 && output_stride1 == 1) {
+    for (int64_t i = 0; i < numel; ++i) {
+      for (int64_t j = 0; j < ddim; ++j) {
+        output_data[add_indices_data[i] * output_stride0 + j] +=
+            src_data[select_indices_data[i] * src_stride0 + j];
+      }
+    }
+  } else {
+    for (int64_t i = 0; i < numel; i++) {
+      THBlas_axpy<T>(ddim, 1,
+              src_data + src_stride0 * select_indices_data[i], src_stride1,
+              output_data + output_stride0 * add_indices_data[i], output_stride1);
+    }
   }
 }
 
+template<>
+void index_select_add<float>(const Tensor &select_indices,
+                             const Tensor &add_indices,
+                             const Tensor &src,
+                             Tensor &output,
+                             const Tensor& offsets) {
+  AT_ASSERT(select_indices.numel() == add_indices.numel());
+  auto add_indices_data = add_indices.data<int64_t>();
+  auto select_indices_data = select_indices.data<int64_t>();
+  auto src_data = src.data<float>();
+  auto output_data = output.data<float>();
+  auto numel = add_indices.numel();
+  int64_t ddim = src.size(1);
+  auto src_stride0 = src.stride(0);
+  auto src_stride1 = src.stride(1);
+  auto output_stride0 = output.stride(0);
+  auto output_stride1 = output.stride(1);
+
+  // We specialize the case where both the src and output are contiguous in the
+  // last dimension to give the compiler a better chance to vectorize.
+  if (src_stride1 == 1 && output_stride1 == 1) {
+    auto accessor = offsets.accessor<int64_t, 1>();
+    std::vector<int> lengths;
+
+    int64_t lower = accessor[0];
+    for (size_t i = 1; i < offsets.numel(); ++i) {
+      lengths.push_back(accessor[i] - lower);
+      lower = accessor[i];
+    }
+    lengths.push_back(numel - lower);
```

Comment:
maybe add a TODO to just changed the underlying kernel to take offsets, not lengths?